BootloaderCorePkg/Library/AcpiInitLib/AcpiInitLib.c

```diff
@@ -345,52 +350,76 @@ UpdateMadt (
   IN UINT8   *Current
   )
 {
+  EFI_ACPI_DESCRIPTION_HEADER                 *Table;
+  UINT8                                       *TempMadt;
+  UINT8                                       *MadtPtr;
+  UINT8                                       *MadtEnd;
+  EFI_ACPI_MADT_ENTRY_COMMON_HEADER           *EntryHeader;
   EFI_ACPI_5_0_PROCESSOR_LOCAL_APIC_STRUCTURE *LocalApic;
-  EFI_ACPI_5_0_MULTIPLE_APIC_DESCRIPTION_TABLE_HEADER *Temp;
-  SYS_CPU_INFO            *SysCpuInfo;
-  UINT32                   CpuCount;
-  UINT32                   MarkCpuCount;
-  UINT32                   Idx;
-  UINT32                   EndOfMadt;
-  UINT8                   *Madt;
-
-  Temp = (EFI_ACPI_5_0_MULTIPLE_APIC_DESCRIPTION_TABLE_HEADER *)Current;
-  EndOfMadt = Temp->Header.Length + (UINT32)Current;
-
-  Madt = Current;
-  //Skip the header
-  Madt = Madt + sizeof (EFI_ACPI_5_0_MULTIPLE_APIC_DESCRIPTION_TABLE_HEADER);
-
-  //Get the Mp Table info
+  SYS_CPU_INFO                                *SysCpuInfo;
+  UINT32                                       Length;
+  UINT32                                       Index;
+
+  //
+  // Get the number of detected CPUs
+  //
   SysCpuInfo = MpGetInfo ();
-  if (SysCpuInfo != NULL) {
-    CpuCount = SysCpuInfo->CpuCount;
-    MarkCpuCount = CpuCount;
-  } else {
-    return EFI_UNSUPPORTED;
+  if ((SysCpuInfo == NULL) || (SysCpuInfo->CpuCount == 0)) {
+    return EFI_ABORTED;
   }
 
-  //Search for Processor Local APIC
-  while ((UINT32)Madt < EndOfMadt) {
-    LocalApic = (EFI_ACPI_5_0_PROCESSOR_LOCAL_APIC_STRUCTURE *)Madt;
-    if (LocalApic->Type == EFI_ACPI_5_0_PROCESSOR_LOCAL_APIC) {
-      LocalApic->Flags = 0;
-      for (Idx = 0 ; Idx < CpuCount; Idx ++) {
-        if ((LocalApic->ApicId == SysCpuInfo->CpuInfo[Idx].ApicId) && (LocalApic->Flags == 0)) {
-          LocalApic->Flags = 1;
-          MarkCpuCount--;
-          break;
-        }
-      }
-    }
-    Madt = Madt + LocalApic->Length;
+  Table = (EFI_ACPI_DESCRIPTION_HEADER *)Current;
+
+  //
+  // Copy original MADT
+  //
+  TempMadt = (UINT8 *)AllocateTemporaryMemory (0);
+  CopyMem ((VOID *)TempMadt, (VOID *)Table, Table->Length);
+
+  MadtPtr = TempMadt;
+  MadtEnd = MadtPtr + Table->Length;
+
+  //
+  // Copy MADT Header
+  //
+  Length = sizeof (EFI_ACPI_5_0_MULTIPLE_APIC_DESCRIPTION_TABLE_HEADER);
+  CopyMem ((VOID *)Current, (VOID *)MadtPtr, Length);
+  MadtPtr += Length;
+  Current += Length;
+
+  //
+  // Append Processor Local APIC info
+  //
+  LocalApic = (EFI_ACPI_5_0_PROCESSOR_LOCAL_APIC_STRUCTURE *)Current;
```

Comment:
Is there any specific reason we want to put LOCAL APIC entries at the top ?
I was thinking it might be easier to copy the original MADT table and then append local APIC entries, then adjust the final length.

Reply:
> Is there any specific reason we want to put LOCAL APIC entries at the top ?
> I was thinking it might be easier to copy the original MADT table and then append local APIC entries, then adjust the final length.

The original MADT has already fixed size Local APIC entries, so this hook ignores existing one. Basically, MADT entries are sorted by its TYPE, and Local APIC type is first item as Type 0. It should be okay to append it to to top or bottom since it's enumerated by Type, but this is just to keep sorted order.

Reply:
Do you think we can just remove the local APIC entries from the original MADT table to make it easier ?

Reply:
maybe save a little bit size, although not too much!  :)

Reply:
Right, if we remove the Local APIC entries from Madt.aslc, we can save the AcpiTable size and this hook would be simpler to append the entries. But, I think better to keep this code, which skips existing Local APIC entries and appends new one, to avoid an issue when Madt.aslc has any Local APIC entry.

Reply:
OK,  I am fine with it.

Reply:
OK,  I am fine with it.